### PR TITLE
Feature/arangosearch levenshtein query

### PR DESCRIPTION
**Pull request by gnusi**
PR adds new AQL functions: `LEVENSHTEIN_MATCH` and 'JACCARD'. 
That's the first stage towards supporting fuzzy search in ArangoSearch, few things are still missing:
* utf8 support
* `LEVENSHTEIN_MATCH` doesn't honor closeness to the supplied target

Both issues will be addressed by upcoming PRs.

### CODE REVIEW

**Comment by gnusi** [comment]
https://github.com/arangodb/planning/issues/2920

File: 3rdParty/iresearch/core/analysis/text_token_stream.cpp [hunk 1]
```diff
@@ -724,7 +724,7 @@ bool make_json_config(
     // explicit_stopwords_set  marks that even empty stopwords list is valid
     rapidjson::Value stopwordsArray(rapidjson::kArrayType);
     if (!options.explicit_stopwords.empty()) {
-      // for simplifying comparison between properties we need deterministic order of stopwords
+      // for simplifying comparsion between properties we need deterministic order of stopwords
```
Comment (Simran-B) [inline 3rdParty/iresearch/core/analysis/text_token_stream.cpp]: This was actually correct
```suggestion
      // for simplifying comparison between properties we need deterministic order of stopwords
```

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8227/

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8227/

File: arangod/Aql/Functions.cpp [hunk 2]
```diff
@@ -4714,6 +4779,61 @@ AqlValue Functions::Intersection(ExpressionContext* expressionContext,
...
+    basics::VelocyPackHelper::VPackHash,
+    basics::VelocyPackHelper::VPackEqual> ValuesMap;
+
+  auto options = trx->transactionContextPtr()->getVPackOptions();
+  ValuesMap values(512, basics::VelocyPackHelper::VPackHash(),
+                        basics::VelocyPackHelper::VPackEqual(options));
+
+  AqlValue const& lhs = extractFunctionParameterValue(args, 0);
+
+  if (!lhs.isArray()) {
```
Comment (Dronplane) [inline arangod/Aql/Functions.cpp]: Maybe wrap with ADB_UNLIKELY ?

File: arangod/Aql/Functions.cpp [hunk 3]
```diff
@@ -4714,6 +4779,61 @@ AqlValue Functions::Intersection(ExpressionContext* expressionContext,
...
+    // not an array
+    registerWarning(ctx, AFN, TRI_ERROR_QUERY_ARRAY_EXPECTED);
+    return AqlValue(AqlValueHintNull());
+  }
+
+  AqlValue const& rhs = extractFunctionParameterValue(args, 1);
+
+  if (!rhs.isArray()) {
```
Comment (Dronplane) [inline arangod/Aql/Functions.cpp]: Maybe wrap with ADB_UNLIKELY ?

File: arangod/IResearch/VelocyPackHelper.h [hunk 4]
```diff
@@ -160,6 +160,23 @@ inline irs::string_ref getStringRef(VPackSlice const& slice) {
...
+///        must be a string
+/// @return extracted string_ref
+//////////////////////////////////////////////////////////////////////////////
+inline irs::bytes_ref getBytesRef(VPackSlice const& slice) {
+  TRI_ASSERT(slice.isString());
+
+  arangodb::velocypack::ValueLength size;
+  auto const* str = slice.getString(size);
+
+  static_assert(sizeof(arangodb::velocypack::ValueLength) == sizeof(size_t),
```
Comment (Dronplane) [inline arangod/IResearch/VelocyPackHelper.h]: Maybe relax check from == to <=

File: tests/IResearch/IResearchFilterFunction-test.cpp [hunk 5]
```diff
@@ -4960,7 +4961,255 @@ TEST_F(IResearchFilterFunctionTest, wildcard) {
...
+          .scored_terms_limit(128)
+          .term("foo");
+
+    assertFilterSuccess(
+        vocbase(),
+        "FOR d IN myView FILTER LEVENSHTEIN_MATCH(d.name, 'foo', 1) RETURN d",
+        expected);
+    assertFilterSuccess(
+        vocbase(),
+        "FOR d IN myView FILTER LEVENSHTEIN_match(d['name'], 'foo', 1) RETURN d",
```
Comment (Dronplane) [inline tests/IResearch/IResearchFilterFunction-test.cpp]:
```suggestion
        "FOR d IN myView FILTER LEVENSHTEIN_MATCH(d['name'], 'foo', 1) RETURN d",
```

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 6]
```diff
@@ -2186,184 +2302,223 @@ arangodb::Result fromFuncStartsWith(irs::boolean_filter* filter, QueryContext co
...
+
+    if (!res.ok()) {
+      return res;
     }
   }
 
+  if (!withTranspositions && maxDistance > MAX_LEVENSHTEIN_DISTANCE) {
+    return {
+      TRI_ERROR_BAD_PARAMETER,
+      "'"s.append(funcName).append("' AQL function: max Levenshtein distance must be in range [0, 4]")
```
Comment (Dronplane) [inline arangod/IResearch/IResearchFilterFactory.cpp]: Maybe use std::to_string(MAX_LEVENSHTEIN_DISTANCE) to denote range bounds ?

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 7]
```diff
@@ -2186,184 +2302,223 @@ arangodb::Result fromFuncStartsWith(irs::boolean_filter* filter, QueryContext co
...
+    };
+  } else if (withTranspositions && maxDistance > MAX_DAMERAU_LEVENSHTEIN_DISTANCE) {
+    return {
+      TRI_ERROR_BAD_PARAMETER,
+      "'"s.append(funcName).append("' AQL function: max Damerau-Levenshtein distance must be in range [0, 3]")
```
Comment (Dronplane) [inline arangod/IResearch/IResearchFilterFactory.cpp]: Maybe use std::to_string(MAX_DAMERAU_LEVENSHTEIN_DISTANCE) to denote range bounds ?

File: tests/IResearch/IResearchFilterFunction-test.cpp (hunk 5 above)
Comment (gnusi) [reply tests/IResearch/IResearchFilterFunction-test.cpp]: that's intentional check :-)

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8236/

File: 3rdParty/iresearch/core/analysis/text_token_stream.cpp (hunk 1 above)
Comment (gnusi) [reply 3rdParty/iresearch/core/analysis/text_token_stream.cpp]: `If transpositions are included, does that make it a Damerau–Levenshtein distance?`
@Simran-B yes, correct

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8252/

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8260/

**Comment by gnusi** [comment]
restarted for windows
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr-windows/2224/

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8263/

**Comment by gnusi** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr-windows/2231/EDITION=enterprise,STORAGE_ENGINE=rocksdb,TEST_SUITE=single,limit=windows&&test/

**Pull request merged**